### DISCUSSION TITLE
Clang-tidy crash report location

### DISCUSSION THREAD

**Discussion by MathiasDeRoover** [microsoft/vscode-cpptools#9705] [Q&A] [answered]
Clang-tidy is consistently crashing when I add the -m32 compiler flag to my configuration. The message in the logs is truncated making it difficult to debug what is happening:
```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
  Stack dump:
  0.	Program arguments: /home/mdo/.vscode/extensions/ms-vscode.cpptools-1.12.0-linux-x64/bin/../LLVM/bin/clang-tidy --checks=*,-modernize-use-trailing-return-type,-cppcoreguidelines-avoid-non-const-global-variables,-llvm-*,-llvmlibc-*,-bugprone-reserved-identifier,-cert-dcl37-c,-cert-dcl51-cpp,-hicpp-signed-bitwise,-misc-unused-parameters,-altera-*,-google-readability-todo,-android-*,-google-readability-avoid-underscore-in-googletest-name,-cppcoreguidelines-avoid-magic-numbers,-cppcoreguidelines-pro-type-vararg,-hicpp-vararg --header-filter=/home/mdo/develop/nhc-embedded --export-fixes=/tmp/loose/cpptools/11132614347462837328/fixes140073689401088.yaml --quiet /home/mdo/develop/nhc-embedded/applications/dalpluginnhc/test/diagnostics_test.cc -- -std=c++20 -xc++ -Wno-pragma-pack -DPROJECT_NAME -D__SSP_STRONG__=3 -D__DBL_MIN_EXP__=(-1021) -D__FLT32X_MAX_…
 ```
 Could somebody point me in the right direction or tell me how to get the full message? 
 Thank you in advance!
PS: my config file for completion sake:
```
{
    "configurations": [
        {
            "name": "LinuxC",
            "includePath": [
                "${workspaceFolder}/**",
                ...
            ],
            "defines": [
                "PROJECT_NAME"
            ],
            "cStandard": "c99",
            "cppStandard": "c++20",
            "browse": {
                "path": [
                    "${workspaceFolder}"
                ]
            },
            "compilerArgs": [
                "-m32"
            ],
            "compilerPath": "/usr/bin/gcc"
        }
    ],
    "version": 4
}
```

**Comment by sean-mcmanus** [comment, marked as answer]
I created issue https://github.com/microsoft/vscode-cpptools/issues/9706 -- the full logging gets shown on Windows, i.e. the logging isn't being handled correctly on Linux/Mac, seemingly due to differences in the stderr/etc. output.

**Reply by MathiasDeRoover** [reply]
@sean-mcmanus I suspect that me using zsh by default causes clang-tidy to crash.
The following command crashes:
```
/home/mdo/.vscode/extensions/ms-vscode.cpptools-1.12.0-linux-x64/bin/../LLVM/bin/clang-tidy --header-filter=/home/mdo/niko/docker-files/develop/nhc-embedded --export-fixes=/tmp/loose/cpptools/11132614347462837328/fixes140275326359296.yaml --quiet /home/mdo/develop/nhc-embedded/applications/dalpluginnhc/test/diagnostics_test.cc -- -std=c++20 -xc++ -Wno-pragma-pack -DPROJECT_NAME -D__SSP_STRONG__=3 -D__DBL_MIN_EXP__=(-1021)
zsh: unknown file attribute: 1
```
While the following command at least doesn't crash:
```
/home/mdo/.vscode/extensions/ms-vscode.cpptools-1.12.0-linux-x64/bin/../LLVM/bin/clang-tidy --header-filter=/home/mdo/develop/nhc-embedded --export-fixes=/tmp/loose/cpptools/11132614347462837328/fixes140275326359296.yaml --quiet /home/mdo/develop/nhc-embedded/applications/dalpluginnhc/test/diagnostics_test.cc -- -std=c++20 -xc++ -Wno-pragma-pack -DPROJECT_NAME -D__SSP_STRONG__=3 -D__DBL_MIN_EXP__="-1021"
2 warnings and 2 errors generated.
Error while processing /home/mdo/develop/nhc-embedded/applications/dalpluginnhc/test/diagnostics_test.cc.
warning: '__DBL_MIN_EXP__' macro redefined [clang-diagnostic-macro-redefined]
note: previous definition is here
/home/mdo/develop/nhc-embedded/applications/dalpluginnhc/test/diagnostics_test.cc:1:10: error: 'gtest/gtest.h' file not found [clang-diagnostic-error]
#include <gtest/gtest.h>
         ^~~~~~~~~~~~~~~
Error opening output file: No such file or directory
```
Any thoughts?

**Reply by sean-mcmanus** [reply]
Well, I fixed the logging when there is a crash for our next release. It's possible we could be contributing to the crash if we're passing unexpected/invalid arguments to it, but if you're reproing the crash via running the command on the command line then it could be a clang-tidy bug. I'm not sure what's going on with the zsh.

**Reply by MathiasDeRoover** [reply]
@sean-mcmanus Another update: I got past the crash by using the system installed clang-tidy instead of the one shipped with the extension, but now I have an issue with the arguments being passed.
My goal is to build for 32bit on a 64bit system so I add the -m32 compiler flag which according to the logs generates the following arguments for clang-tidy:
[clang_tidy_arguments.txt](https://github.com/microsoft/vscode-cpptools/files/9299026/clang_tidy_arguments.txt)
However I noticed that clang-tidy adds implicit defines like '-D__x86_64__'. If I feed the generated arguments directly into clang it also generates those implicit defines. However if I add the '-m32' flag to the clang-tidy (clang) arguments in addition to the generated ones, it does produce the defines that I expect.
Any thoughts?

**Reply by sean-mcmanus** [reply]
Yeah, I see -- it works for me if I set
```json
    "C_Cpp.codeAnalysis.clangTidy.args": [
        "--extra-arg=-m32"
    ]
```
 is that what you tried? I'm not sure if we should change it to automatically add compilerArgs (or a subset of them) to --extra-arg or not.

**Reply by MathiasDeRoover** [reply]
Ahh, I haven't tried that yet no. I'll try it tomorrow but I believe it should work. Thanks for the tip, this approach is sufficient for me.
